Layer order now preserved on refresh or on load from URL or JSON server

manuel-castro opened the pull request:
This PR changes the "layers" property in the JSON state from an object to an array to preserve the layer order (see https://github.com/seung-lab/neuroglancer/issues/228).

manuel-castro commented:
I signed it!

jbms commented on src/neuroglancer/layer_specification.ts:
```diff
@@ -127,10 +128,22 @@ export class TopLevelLayerListSpecification extends RefCounted implements LayerL
+      // If array, layers have an order
+      for (const layerObj of x) {
+        verifyObject(layerObj);
+        const name = layerObj.name;
```
```suggestion
        const name = verifyObjectProperty(layerObj, 'name', verifyString);
```

jbms commented:
This looks good, but can you also fix the Python code to support this format as well?

manuel-castro commented:
My mistake. Thanks for the feedback

jbms commented:
Thanks --- ManagedLayer.to_json in viewer_state.py also needs to be fixed to include the name property now

Since there aren't unit tests for this stuff at the moment, please make sure to manually test that e.g. the examples in python/neuroglancer/examples still work.

manuel-castro commented:
Sorry about that again, I had difficulty running some of the Python code, but I figured out what the problem was and was able to. I added the name property to ManagedLayer.to_json, removed the deletion of the name property in TypeScript for symmetry, and changed what I believe was a typo in one of the examples.

jbms commented on src/neuroglancer/layer_specification.ts:
```diff
@@ -127,10 +128,20 @@ export class TopLevelLayerListSpecification extends RefCounted implements LayerL
+      // If array, layers have an order
+      for (const layerObj of x) {
+        verifyObject(layerObj);
+        const name = verifyObjectProperty(layerObj, 'name', verifyString);
```
```suggestion
        const name = this.layerManager.getUniqueLayerName(verifyObjectProperty(layerObj, 'name', verifyString));
```
